### PR TITLE
Adds LayoutService interface to allow native layout updates

### LINKED ISSUE #9988: Components that require native layout don't have a way to trigger a yoga re-layout
We need components that have `RequireNativeLayout()` return true be able to trigger a re-layout.
@acoates-ms:
> ...wrap the element in a custom panel, which knows when measure is called on itself.

When adding a view to the xaml tree that says RequireNativeLayout() - but its parent doesn't, add a custom panel in between that overrides MeasureInvalidated to cause yoga to re-layout

### PR DESCRIPTION
**Pull request by rozele**
### Type of Change
_Erase all that don't apply._
- New feature (non-breaking change which adds functionality)

### Why
In order to enable native layout updates, we need to expose an interface to mark Yoga nodes as dirty, and also to trigger native layout.

Resolves #9988
Towards #7601
Workaround for #9343

### What
While we could just blindly invoke `onBatchComplete` on the NativeUIManager, it is safer to also expose an interface to test whether the UIManager is currently processing a batch of UIManager operations so we can piggy back on the current onBatchComplete call that will occur at the end of a batch if one is in progress. This is also likely to prevent UI tearing. Today, it is unlikely that an call from the UI thread to invoke Yoga and XAML layout will interrupt a batch, if we eventually implement something like eager view creation, this may no longer be the case.

Also, rather than calling `onBatchComplete`, which may have the side-effect of calling UpdateLayout on nodes that specify `NeedsForceLayout`, native UI updates will not involve the creation of new nodes, so we can skip this step, thus the addition of a new API to call YGNodeCalculateLayout for a specific root (and skip layout for other roots where it may not be needed). The LayoutService can also conditionally apply layout to all roots, a specific root with unbound constraints, or a specific root with known constraints.

We will be using this service for a number of features, including re-entrant Yoga layout for views that are self-measuring with React Native descendants dependent on Yoga layout, as well as to work around a text truncation bug caused by Yoga caching to force the Text to re-measure and re-layout when truncation is detected (is narrow use cases).

## Testing
We have used this new LayoutService API in Messenger to implement a native module that dirties some text nodes that get truncated due to a Yoga bug with cached flex basis values. There is no change to core behaviors, I also tested that layout still works correctly in Playground / RNTester.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/10671)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/LayoutService.idl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    void ApplyLayout(Int64 reactTag);
+
+    DOC_STRING("Applies layout from a given React node with root layout constraints.")
+    void ApplyLayout(Int64 reactTag, Single width, Single height);
+
+    DOC_STRING(
+      "Determines whether the UIManager is currently processing a batch of node updates."
+      "This is useful for optimizing layout and ensuring that applying layout to a particular "
+      "node will not cause tearing in the rendered UI.")
+    Boolean IsInBatch();
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/LayoutService.idl:31]: this sounds like a get-only property?

File: vnext/Microsoft.ReactNative/LayoutService.idl [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  [webhosthidden]
+  DOC_STRING("Provides access to Yoga layout functionality.")
+  runtimeclass LayoutService {
+    DOC_STRING("Use this method to get access to the @LayoutService associated with the @IReactContext.")
+    static LayoutService FromContext(IReactContext context);
+
+    DOC_STRING("Applies layout to all root nodes.")
+    void ApplyLayout();
+
+    DOC_STRING("Applies layout from a given React node.")
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/LayoutService.idl:21]: these operations are recursive, I reckon? if so, worth clarifying in the doc string?

File: vnext/Microsoft.ReactNative/LayoutService.idl (hunk 1 above)
Comment (rozele) [reply vnext/Microsoft.ReactNative/LayoutService.idl:31]: Good point. I can switch it to a property :)

File: vnext/Microsoft.ReactNative/LayoutService.idl (hunk 2 above)
Comment (rozele) [reply vnext/Microsoft.ReactNative/LayoutService.idl:21]: And I think I'll eliminate the unconstrained ApplyLayout overload, I was adding that for the future ability to nest views inside text, but that is an internal API and doesn't need an ABI interface.

**Review by asklar — APPROVED** [review]

**Comment by rozele** [comment]
Not convinced the CI failure is a deterministic failure, can a maintainer please nudge ADO to re-run the failing test?

**Comment by asklar** [comment]
nudged

**Review by jonthysell — APPROVED** [review]

**Comment by rozele** [comment]
> Is this something you might be able to add @rozele ?

@asklar - agree it's a good idea, but the last time I went down the rabbit hole of trying to create a shared native module project for Visual Studio that could be used across both the Playground and Playground-win32 solution, my very much outdated MSBuild skills failed me. I think it would be a great idea to generally have a Playground native modules project where various view manager and native module APIs are exercised, but there isn't a precedent yet. If I get some time again in the near future, I'll try to set one up, but would be much obliged if someone knows of a pre-existing template or script we can run to generate a native module project for the Playground solutions.

**Comment by asklar** [comment]
> > Is this something you might be able to add @rozele ?
> 
> @asklar Alexander Sklar FTE - agree it's a good idea, but the last time I went down the rabbit hole of trying to create a shared native module project for Visual Studio that could be used across both the Playground and Playground-win32 solution, my very much outdated MSBuild skills failed me. I think it would be a great idea to generally have a Playground native modules project where various view manager and native module APIs are exercised, but there isn't a precedent yet. If I get some time again in the near future, I'll try to set one up, but would be much obliged if someone knows of a pre-existing template or script we can run to generate a native module project for the Playground solutions.
> 
> I generally agree that it's fair to hold off on merging this PR until we have some code that exercises this functionality (and perhaps an e2e test).

@rozele we have samples both in the samples repo (those are manually upgraded and are more e2e apps) and here: https://github.com/microsoft/react-native-windows/tree/main/packages/sample-apps/windows/SampleLibraryCPP
would augmenting that one not suffice?

**Comment by chiaramooney** [comment]
@rozele is this ready to merge? Or is the plan to add a sample within this PR?

**Comment by rozele** [comment]
There's some overhead involved in creating a sample, because we do not have any custom view managers or native modules in this repo for Playground (only those consumed via NPM), so I do plan to get around to this eventually, just not right away.

**Pull request merged**